source/fuzz/transformation_propagate_instruction_up.cpp

```diff
@@ -0,0 +1,393 @@
+    op_phi_operands.push_back({SPV_OPERAND_TYPE_ID, {new_result_id}});
+    op_phi_operands.push_back({SPV_OPERAND_TYPE_ID, {predecessor_id}});
+
+    // Create a clone of the |inst| to be inserted into the |predecessor_id|.
+    std::unique_ptr<opt::Instruction> clone(inst->Clone(ir_context));
+    clone->SetResultId(new_result_id);
+
+    fuzzerutil::UpdateModuleIdBound(ir_context, new_result_id);
+
+    // Adjust |clone|'s operands to account for possible dependencies on OpPhi
+    // instructions from the same basic block.
+    for (uint32_t i = 0; i < clone->NumInOperands(); ++i) {
+      auto& operand = clone->GetInOperand(i);
+
+      if (operand.type == SPV_OPERAND_TYPE_ID &&
+          op_phi_to_result_id.count(operand.words[0])) {
+        operand.words[0] =
+            op_phi_to_result_id.at(operand.words[0]).at(predecessor_id);
+      }
+    }
+
+    // Insert cloned instruction into the predecessor.
+    auto* predecessor = ir_context->cfg()->block(predecessor_id);
+    assert(predecessor && "|predecessor_id| is invalid");
+
+    GetLastInsertBeforeInstruction(predecessor, clone->opcode())
+        ->InsertBefore(std::move(clone));
+  }
+
+  // Insert an OpPhi instruction into the basic block of |inst|.
+  ir_context->get_instr_block(inst)->begin()->InsertBefore(
+      MakeUnique<opt::Instruction>(ir_context, SpvOpPhi, inst->type_id(),
+                                   inst->result_id(),
+                                   std::move(op_phi_operands)));
+
+  // Remove |inst| from the basic block.
+  inst->RemoveFromList();
+  delete inst;  // RemoveFromList doesn't clear the memory.
+
+  // Make sure our changes are analyzed
+  ir_context->InvalidateAnalysesExceptFor(
+      opt::IRContext::Analysis::kAnalysisNone);
+
+  // Remove unused OpPhi instructions.
+  for (const auto& entry : op_phi_to_result_id) {
+    auto op_phi_result_id = entry.first;
+
+    if (ir_context->get_def_use_mgr()->NumUsers(op_phi_result_id) == 0) {
+      auto* unused_op_phi_inst =
+          ir_context->get_def_use_mgr()->GetDef(op_phi_result_id);
+      assert(unused_op_phi_inst && "|op_phi_result_id| must be valid");
+
+      unused_op_phi_inst->RemoveFromList();
+      delete unused_op_phi_inst;
+    }
+  }
+
+  // Make sure our changes are analyzed after we've removed unused OpPhi
+  // instructions.
+  ir_context->InvalidateAnalysesExceptFor(
+      opt::IRContext::Analysis::kAnalysisNone);
+}
+
+protobufs::Transformation TransformationPropagateInstructionUp::ToMessage()
+    const {
+  protobufs::Transformation result;
+  *result.mutable_propagate_instruction_up() = message_;
+  return result;
+}
+
+bool TransformationPropagateInstructionUp::IsOpcodeSupported(SpvOp opcode) {
+  // TODO(https://github.com/KhronosGroup/SPIRV-Tools/issues/3605):
+  //  We only support "simple" instructions that don't work with memory.
+  //  We should extend this so that we support the ones that modify the memory
+  //  too.
```

Comment:
How about we right away support operations that read memory - i.e., load instructions - and just skip over instructions that modify memory, as well as barriers?

I don't see any problem moving loads up into the previous block, skipping over OpPhis.

Reply:
Adding support for memory instructions (including those that read memory) can be tricky. The tricky part comes from memory barriers. Consider the following example.
```
   [1:]
 /      \
|    [2: OpMemoryBarrier]
 \      /
 [3: OpLoad]
```
If we try to move `OpLoad` into `1` block, then (as far as my understanding goes) we might violate the programme's semantics since `OpLoad` can be executed before `OpMemoryBarrier`. Thus, I would prefer to defer the implementation of memory instructions. I might be missing something here, though. @afd, what do you think?

Reply:
We've decided to add support for memory instructions in a separate PR.